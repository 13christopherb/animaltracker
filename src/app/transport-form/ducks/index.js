export {actions as transportFormActions} from './actions';
export {types as transportFormTypes} from './types';