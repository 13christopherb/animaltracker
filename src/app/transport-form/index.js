export {TransportForm} from './transport-form';
export {TransportFormContainer} from './transport-form-container';
export {TransportModalForm} from './transport-modal-form';