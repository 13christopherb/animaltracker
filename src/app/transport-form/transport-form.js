import React from 'react';
import moment from 'moment';
import {Select, Input} from "../form-fields/index";
import {Button, Col, Form} from 'react-bootstrap';
import {Field} from 'redux-form'

export const TransportForm = ({handleSubmit, handleChange, onSubmit, ...props}) => {
    return (
        <Form onSubmit={handleSubmit(onSubmit)}>
            <Form.Row>
                <Col md={2}>
                    <Field
                        component={Select}
                        title="Departs"
                        id="departs"
                        name="departs"
                        value={props.value}
                        onChange={handleChange}
                        options={[
                            'NRO',
                            'MBO',
                            'SLO'
                        ]}
                    />
                </Col>
                <Col md={2}>
                    <Field
                        component={Select}
                        title="Arrives"
                        name="arrives"
                        id="arrives"
                        value={props.value}
                        onChange={handleChange}
                        options={[
                            'NRO',
                            'MBO',
                            'SLO'
                        ]}
                    />
                </Col>
                <Col md={2}>
                    <Field
                        component={Input}
                        name="meetPlace"
                        id="meetPlace"
                        title="Meet Place"
                        type="text"
                        value={props.value}
                        onChange={handleChange}
                    />
                </Col>
                <Col md={2}>
                    <Field
                        component={Select}
                        name="meetDate"
                        id="meetDate"
                        title="Meet Date"
                        value={props.value}
                        onChange={handleChange}
                        options={[
                            moment().format('dddd MM/DD'),
                            moment().add(1, 'days').format('dddd MM/DD'),
                            moment().add(2, 'days').format('dddd MM/DD'),
                            moment().add(3, 'days').format('dddd MM/DD'),
                        ]}
                    />
                </Col>
                <Col md={2}>
                    <Field
                        component={Input}
                        name="meetTime"
                        id="meetTime"
                        title="Meet Time"
                        type="text"
                        value={props.value}
                        onChange={handleChange}
                    />
                </Col>
                <Col md={3}>
                    <Button variant="primary"
                            disabled={props.pristine || !props.valid}
                            type="submit">
                        Submit
                    </Button>
                </Col>
            </Form.Row>
        </Form>
    )
};