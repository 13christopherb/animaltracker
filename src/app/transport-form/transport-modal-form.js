import React from 'react';
import {Field} from 'redux-form'
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {Button, Col, Form, Modal} from 'react-bootstrap';
import {Input, Select} from "../form-fields/index";
import moment from "moment";

/**
 * Form for creating a new animal in the form of a modal
 * @param handleSubmit
 * @param handleChange
 * @param onSubmit
 * @param props
 * @returns {*}
 * @constructor
 */
export const TransportModalForm = ({handleSubmit, handleChange, onSubmit, toggleModal, show, ...props}) => {
    return (
        <Modal show={show} onHide={toggleModal}>
            <Modal.Header closeButton>
                <Modal.Title>New Transport</Modal.Title>
            </Modal.Header>
            <Form onSubmit={handleSubmit(onSubmit)}>
                <Modal.Body className="modal-body">
                    <Form.Row>
                        <Col xs={6}>
                            <Field
                                component={Select}
                                title="Departs"
                                id="departs"
                                name="departs"
                                value={props.value}
                                onChange={handleChange}
                                options={[
                                    'NRO',
                                    'MBO',
                                    'SLO'
                                ]}
                            />
                        </Col>
                        <Col xs={6}>
                            <Field
                                component={Select}
                                title="Arrives"
                                name="arrives"
                                id="arrives"
                                value={props.value}
                                onChange={handleChange}
                                options={[
                                    'NRO',
                                    'MBO',
                                    'SLO'
                                ]}
                            />
                        </Col>
                    </Form.Row>
                    <Form.Row>
                        <Col xs={12}>
                            <Field
                                component={Input}
                                name="meetPlace"
                                id="meetPlace"
                                title="Meet Place"
                                type="text"
                                value={props.value}
                                onChange={handleChange}
                            />
                        </Col>
                    </Form.Row>
                    <Form.Row>
                        <Col xs={8}>
                            <Field
                                component={Select}
                                name="meetDate"
                                id="meetDate"
                                title="Meet Date"
                                value={props.value}
                                onChange={handleChange}
                                options={[
                                    moment().format('dddd MM/DD'),
                                    moment().add(1, 'days').format('dddd MM/DD'),
                                    moment().add(2, 'days').format('dddd MM/DD'),
                                    moment().add(3, 'days').format('dddd MM/DD'),
                                ]}
                            />
                        </Col>
                    </Form.Row>
                    <Form.Row>
                        <Col xs={6}>
                            <Field
                                component={Input}
                                name="meetTime"
                                id="meetTime"
                                title="Meet Time"
                                type="text"
                                value={props.value}
                                onChange={handleChange}
                            />
                        </Col>
                    </Form.Row>
                </Modal.Body>
                <Modal.Footer>
                    <Button variant="primary" onClick={toggleModal}
                            disabled={props.pristine || !props.valid}
                            type="submit">
                        Submit
                    </Button>
                    <Button variant="danger" onClick={toggleModal}
                            type="button">
                        Cancel
                    </Button>
                </Modal.Footer>
            </Form>
        </Modal>
    );
};