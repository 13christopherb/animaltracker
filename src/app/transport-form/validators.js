import moment from 'moment';

/**
 * Contains validators for TransportForm component
 */

/**
 * Validation function for required values for provided fields
 * @param values Field values to be validated
 */
export const validate = values => {
    const errors = {};
    if (!values.meetTime) {
        errors.meetTime = 'Required';
    } else if (!/^([0-1]?\d|(?:2[0-3])):?([0-5]\d)$/.test(values.meetTime)) {
        errors.meetTime = 'Must be a valid military time'
    }
    if (!values.meetPlace) {
        errors.meetPlace = 'Required';
    } else if (!/^[a-zA-Z\s]+$/.test(values.meetPlace)) {
        errors.meetPlace = 'Must only contain letters';
    } else if (values.meetPlace.length > 20) {
        errors.meetPlace = 'Must be fewer than 20 characters';
    }
    if (values.departs === values.arrives) {
        errors.arrives = 'Cannot be the same as Departs';
    }
    return errors
};