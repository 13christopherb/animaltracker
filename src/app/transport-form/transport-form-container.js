import React, {Component} from 'react';
import {connect} from 'react-redux';
import moment from 'moment';
import {transportFormActions} from "./ducks";
import {TransportForm} from "./transport-form";

import {reduxForm} from "redux-form";
import {validate} from './validators';

class TransportFormContainer extends Component {

    constructor(props) {
        super(props);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit(values) {
        let meetDateTime = moment(values.meetDate + ' ' + values.meetTime,'dddd MM/DD h:mm' );
        const transport = {
            departs: values.departs,
            arrives: values.arrives,
            meetPlace: values.meetPlace,
            meetTime: meetDateTime.format()
        };
        this.props.dispatch(transportFormActions.addTransport(transport));
    }

    render() {
        return (
            React.cloneElement(this.props.children, {onSubmit: this.handleSubmit, ...this.props})
        );
    }
}

function mapStateToProps({}, ownProps) {
    return {
        defaultTime: ownProps.defaultTime
    }
}

const connected = reduxForm({
    form: 'TransportForm',
    validate
})(connect(mapStateToProps)(TransportFormContainer));

export {connected as TransportFormContainer};