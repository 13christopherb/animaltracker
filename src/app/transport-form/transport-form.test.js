import React from 'react';
import {shallow} from 'enzyme';
import {TransportForm} from "./transport-form";
import toJson from 'enzyme-to-json';
import moment from 'moment';

const transport = {
    departs: 'MBO',
    arrives: 'NRO',
    meetTime: moment('2017-09-15 09:30:00') //Arbitrary time
};

describe('<TransportForm />', () => {
    describe('render()', () => {
        test('should render the component', () => {
            const mockSubmit = jest.fn();
            const mockChange = jest.fn();
            const mockDateTimeChange = jest.fn();
            const defaultTime = moment('2017-09-15 09:30:00') //Arbitrary time
            const wrapper = shallow(<TransportForm handleSubmit={mockSubmit}
                                                   handleInputChange={mockChange}
                                                   handleDateTimeChange={mockDateTimeChange}
                                                   arrives={transport.arrives}
                                                   departs={transport.departs}
                                                   meetTime={transport.meetTime}
                                                   defaultTime={defaultTime}/>);

            expect(toJson(wrapper)).toMatchSnapshot()
        })
    });
});