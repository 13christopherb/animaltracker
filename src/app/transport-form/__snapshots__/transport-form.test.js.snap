// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransportForm /> render() should render the component 1`] = `
<div>
  <form>
    <div
      className="form-row"
    >
      <div
        className="col col-md-2"
      >
        <Field
          component={[Function]}
          id="departs"
          name="departs"
          options={
            Array [
              "NRO",
              "MBO",
              "SLO",
            ]
          }
          title="Departs"
        />
      </div>
      <div
        className="col col-md-2"
      >
        <Field
          component={[Function]}
          id="arrives"
          name="arrives"
          options={
            Array [
              "NRO",
              "MBO",
              "SLO",
            ]
          }
          title="Arrives"
        />
      </div>
      <div
        className="col col-md-2"
      >
        <Field
          component={[Function]}
          id="meetDate"
          name="meetDate"
          options={
            Array [
              "Thursday 11/08",
              "Friday 11/09",
              "Saturday 11/10",
              "Sunday 11/11",
            ]
          }
          title="Meet Date"
        />
      </div>
      <div
        className="col col-md-2"
      >
        <Field
          component={[Function]}
          id="meetTime"
          name="meetTime"
          title="Meet Time"
          type="text"
        />
      </div>
      <div
        className="col pt-4 col-md-3"
      >
        <button
          className="btn btn-primary"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</div>
`;
